lib/db_connection.ex

```diff
@@ -999,87 +989,61 @@ defmodule DBConnection do
     end
   end
 
-  defp checkin(conn) do
-    case delete_info(conn) do
-      {:ok, conn_state} ->
-        %DBConnection{holder: holder, pool_ref: pool_ref} = conn
-        _ = holder.checkin(pool_ref, conn_state)
-        :ok
-      {:error, _} = error ->
-        error
-    end
+  defp checkin(%DBConnection{pool_ref: pool_ref}) do
+    Holder.checkin(pool_ref)
   end
 
   defp checkin(%DBConnection{} = conn, fun, meter, opts) do
-    with {:ok, conn_state, meter} <- fetch_info(conn, meter) do
-      return = fun.(conn, conn_state, meter, opts)
-      checkin(conn)
-      return
-    end
+    return = fun.(conn, meter, opts)
+    checkin(conn)
+    return
   end
   defp checkin(pool, fun, meter, opts) do
     run(pool, fun, meter, opts)
   end
 
-  defp delete_disconnect(conn, conn_state, err) do
-    _ = delete_info(conn)
-    %DBConnection{holder: holder, pool_ref: pool_ref} = conn
-    _ = holder.disconnect(pool_ref, err, conn_state)
+  defp disconnect(%DBConnection{pool_ref: pool_ref}, err) do
+    _ = Holder.disconnect(pool_ref, err)
     :ok
   end
 
-  defp delete_stop(conn, conn_state, kind, reason, stack) do
-    _ = delete_info(conn)
-    msg = "client #{inspect self()} stopped: " <>
-      Exception.format(kind, reason, stack)
+  defp stop(%DBConnection{pool_ref: pool_ref}, kind, reason, stack) do
+    msg = "client #{inspect self()} stopped: " <> Exception.format(kind, reason, stack)
     exception = DBConnection.ConnectionError.exception(msg)
-    %DBConnection{holder: holder, pool_ref: pool_ref} = conn
-    _ = holder.stop(pool_ref, exception, conn_state)
+    _ = Holder.stop(pool_ref, exception)
     :ok
   end
 
-  defp handle(conn, conn_state, fun, args, opts) do
-    %DBConnection{conn_mod: conn_mod} = conn
-
-    try do
-      apply(conn_mod, fun, args ++ [opts, conn_state])
-    catch
-      kind, reason ->
-        stack = System.stacktrace()
-        delete_stop(conn, conn_state, kind, reason, stack)
-        {:catch, kind, reason, stack}
-    end
-  end
-
-  defp handle_common_result(return, conn, conn_state, meter) do
+  defp handle_common_result(return, conn, meter) do
     case return do
-      {:ok, result, conn_state} ->
-        put_info(conn, conn_state)
+      {:ok, result, _conn_state} ->
         {:ok, result, meter}
 
-      {:error, err, conn_state} ->
-        put_info(conn, conn_state)
+      {:error, err, _conn_state} ->
         {:error, err, meter}
 
-      {:disconnect, err, conn_state} ->
-        delete_disconnect(conn, conn_state, err)
+      {:disconnect, err, _conn_state} ->
+        disconnect(conn, err)
         {:error, err, meter}
 
       {:catch, kind, reason, stack} ->
+        stop(conn, kind, reason, stack)
         {kind, reason, stack, meter}
 
       other ->
-        bad_return!(other, conn, conn_state, meter)
+        bad_return!(other, conn, meter)
     end
   end
 
-  defp bad_return!(other, conn, conn_state, meter) do
+  @compile {:inline, bad_return!: 3}
```

Comment:
What stack does this give us in the logs? Does inlining keep the calling function in the stack even when `bad_return!` is called in the tail?

Reply:
>  Does inlining keep the calling function in the stack even when bad_return! is called in the tail?

It does. This is because there may have been cases were `bad_return!` was called outside of the tail and this keeps it more future proof.